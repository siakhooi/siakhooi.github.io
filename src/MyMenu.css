div.mylinkgroup {
  column-count: 4;
  column-fill: balance;
}

div.mylinkgroup  ul {
  margin: 0;
}

button[currentTab="Y"]{
  color:white;
}
button[currentTab="N"]{
  color: #dddddd;
}