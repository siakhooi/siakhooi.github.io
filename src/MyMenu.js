import React, { useState } from 'react';
import ReactDOM from 'react-dom';


function MyMenuBody(props) {
    return (
        <div class="MyMenuBody">{props.children}</div>
    );
}
function MyMenuGroup(props) {
    return (
        <div class="MyMenuGroup">{props.children}</div>
    );
}
function MyMenuGroupCard(props) {
    return (
        <div class="MyMenuGroupCard">{props.children}</div>
    );
}
function MyMenuGroupHeader(props) {
    return (
        <h2 class="MyMenuGroupHeader">{props.children}</h2>
    );
}
function MyMenuHeaders(props) {
    return (
        <div class="MyMenuHeaders">{props.children}</div>
    );
}
function MyMenuHeader(props) {
    return (
        <button class="MyMenuHeader" currentTab={props.selected} onClick={props.onClick}>{props.name}</button>
    );
}
function MyLinkGroup(props) {
    return (
        <div class='MyLinkGroup'>
            <ul class="MyLinkGroup">{props.children}</ul>
        </div>
    );
}
function MyLink(props) {
    let d = props.desc;
    if (d !== undefined) d = `- ${d}`;
    return (
        <li class="MyLink"><a class="MyLink" href={props.url} rel='nofollow'>{props.name}</a>{d}</li>
    );
}

function MyMenu(props) {
    const [displayTab, setDisplayTab] = useState(0);
    let headers = [];
    let body = [];

    props.menu.menu.forEach((x, i) => {
        let selected = (i === displayTab) ? 'Y' : 'N';
        headers.push(<MyMenuHeader id={i} name={x.name} selected={selected} onClick={() => setDisplayTab(i)}></MyMenuHeader>);
        if (i === displayTab) {
            x.menu.forEach((x, i) => {
                let y = x.menu.map((x, i) => {
                    return <MyLink url={x.url} desc={x.description} name={x.name} />
                });
                body.push(
                    <MyMenuGroup>
                        <MyMenuGroupCard>
                            <MyMenuGroupHeader>{x.name}</MyMenuGroupHeader>
                            <MyLinkGroup>{y}</MyLinkGroup>
                        </MyMenuGroupCard>
                    </MyMenuGroup>
                );
            });
        }
    });
    return <div class="MyMenu">
        <MyMenuHeaders>{headers}</MyMenuHeaders>
        <MyMenuBody>{body}</MyMenuBody>
    </div>;
}
function renderMenu(menu, id) {
    return new Promise((resolve, reject) => {
        ReactDOM.render(
            <MyMenu menu={menu} />,
            document.getElementById(id)
        );
        resolve();
    });
}
function renderMyMenu(url, id) {
    fetch(url)
        .then(response => response.json())
        .then(menu => { return renderMenu(menu, id) });
}

export { renderMyMenu };
export default renderMyMenu;
