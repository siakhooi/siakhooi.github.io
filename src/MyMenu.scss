@import "w3";
@import "w3-theme";

div.MyMenu {
  @extend .w3-container;
}
div.MyMenuHeaders {
  @extend .w3-container;
  @extend .w3-theme-d3;
}
button.MyMenuHeader {
  @extend .w3-bar-item;
  @extend .w3-button;
}
button.MyMenuHeader[currentTab="Y"] {
  @extend .w3-theme;
}

div.MyMenuBody {
  @extend .w3-container;
  margin: 0px;
}
div.MyMenuGroup {
  transition: 0.3s;
  display: inline-block;
  vertical-align: top;
  padding: 4px;
  @media screen and (max-width: 576px) {
    width: 100%;
  }
  @media screen and (max-width: 864px) and (min-width: 576px) {
    width: 50%;
  }
  @media screen and (max-width: 1152px) and (min-width: 864px) {
    width: 33%;
  }
  @media screen and (max-width: 1440px) and (min-width: 1152px) {
    width: 25%;
  }
  @media screen and (min-width: 1440px) {
    width: 20%;
  }
}
div.MyMenuGroupCard {
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
  overflow: hidden;
}

h2.MyMenuGroupHeader {
  @extend .w3-bar;
  @extend .w3-theme;
}
//div.MyLinkGroup {}
//ul.MyLinkGroup {}
//li.MyLink { }
//a.MyLink { }
