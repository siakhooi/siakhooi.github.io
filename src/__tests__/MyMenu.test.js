import React from 'react'
import renderer from 'react-test-renderer'
import MyMenu from '../MyMenu'
import fs from 'fs'

it('MyMenu/1', () => {
  const menu = JSON.parse(fs.readFileSync('./src/__tests__/sample.json', { encoding: 'utf8' }))

  const component =
    renderer.create(
      <MyMenu menu={menu} />
    )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
