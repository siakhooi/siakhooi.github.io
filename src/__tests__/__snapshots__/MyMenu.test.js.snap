// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyMenu/1 1`] = `
<div
  className="MyMenu"
>
  <div
    className="MyMenuHeaders"
  >
    <div
      className="MyMenuResponsiveNav"
    >
      <span
        className="MyMenuTitle"
      >
        General
      </span>
      <button
        className="MyMenuBars fa fa-bars"
        onClick={[Function]}
      />
    </div>
    <div
      className="MyMenuNav"
      displayNav="N"
    >
      <button
        className="MyMenuHeader"
        currentTab="Y"
        onClick={[Function]}
      >
        General
      </button>
      <button
        className="MyMenuHeader"
        currentTab="N"
        onClick={[Function]}
      >
        Web
      </button>
      <button
        className="MyMenuHeader"
        currentTab="N"
        onClick={[Function]}
      >
        Java Script
      </button>
      <button
        className="MyMenuHeader"
        currentTab="N"
        onClick={[Function]}
      >
        Cloud
      </button>
    </div>
  </div>
  <div
    className="MyMenuBody"
  >
    <div
      className="MyMenuGroup"
    >
      <div
        className="MyMenuGroupCard"
      >
        <h2
          className="MyMenuGroupHeader"
        >
          References
        </h2>
        <div
          className="MyLinkGroup"
        >
          <ul
            className="MyLinkGroup"
          >
            <li
              className="MyLink"
            >
              <a
                className="MyLink"
                href="https://opensource.org/licenses/alphabetical"
                rel="nofollow"
              >
                OpenSource.org/Licenses
              </a>
            </li>
            <li
              className="MyLink"
            >
              <a
                className="MyLink"
                href="https://spdx.org/licenses/"
                rel="nofollow"
              >
                SPDX.org/Licenses
              </a>
            </li>
            <li
              className="MyLink"
            >
              <a
                className="MyLink"
                href="https://semver.org"
                rel="nofollow"
              >
                Semantic Versioning 2.0.0
              </a>
            </li>
            <li
              className="MyLink"
            >
              <a
                className="MyLink"
                href="https://keepachangelog.com"
                rel="nofollow"
              >
                KeepAChangeLog.com
              </a>
            </li>
            <li
              className="MyLink"
            >
              <a
                className="MyLink"
                href="https://agilemanifesto.org"
                rel="nofollow"
              >
                AgileManifesto.org
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      className="MyMenuGroup"
    >
      <div
        className="MyMenuGroupCard"
      >
        <h2
          className="MyMenuGroupHeader"
        >
          IDE
        </h2>
        <div
          className="MyLinkGroup"
        >
          <ul
            className="MyLinkGroup"
          >
            <li
              className="MyLink"
            >
              <a
                className="MyLink"
                href="https://code.visualstudio.com"
                rel="nofollow"
              >
                VS Code
              </a>
            </li>
            <li
              className="MyLink"
            >
              <a
                className="MyLink"
                href="https://vscodethemes.com"
                rel="nofollow"
              >
                VSCodeThemes.com
              </a>
            </li>
            <li
              className="MyLink"
            >
              <a
                className="MyLink"
                href="https://notepad-plus-plus.org"
                rel="nofollow"
              >
                Notepad++
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <div
      className="MyMenuGroup"
    >
      <div
        className="MyMenuGroupCard"
      >
        <h2
          className="MyMenuGroupHeader"
        >
          Rest Tools
        </h2>
        <div
          className="MyLinkGroup"
        >
          <ul
            className="MyLinkGroup"
          >
            <li
              className="MyLink"
            >
              <a
                className="MyLink"
                href="https://insomnia.rest"
                rel="nofollow"
              >
                Insomnia
              </a>
            </li>
            <li
              className="MyLink"
            >
              <a
                className="MyLink"
                href="https://www.postman.com"
                rel="nofollow"
              >
                Postman
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
